################################################################################
SHORT DESCRIPTION: 
################################################################################
Add cmdline option "delay_init" to force acpi_video_register early on even if
intel opregion is detected on the system.

################################################################################
LONG DESCRIPTION: 
################################################################################
This sounds dangerous at best since for old hardware (sandy-bridge and before?)
with an acpi opregion, acpi_video_register() is explicitely by-passed in
acpi_video_init().

################################################################################
CHANGELOG 
################################################################################
Port to 3.18: Eric Chanudet <chanudete@ainfosec.com>
Port to 4.14: Richard Turner <turnerr@ainfosec.com>

################################################################################
REMOVAL 
################################################################################
Recommended for removal unless we figure out why this is in here.

################################################################################
UPSTREAM PLAN
################################################################################
None, upstream clearly does something different.

################################################################################
INTERNAL DEPENDENCIES 
################################################################################
xenclient-dom0-tweak adds delay_init on grub's kernel cmdline.

################################################################################
PATCHES 
################################################################################
--- a/drivers/acpi/acpi_video.c
+++ b/drivers/acpi/acpi_video.c
@@ -83,6 +83,9 @@ module_param(device_id_scheme, bool, 044
 static int only_lcd = -1;
 module_param(only_lcd, int, 0444);
 
+static bool delay_init = 0;
+module_param(delay_init, bool, 0644);
+
 static int register_count;
 static DEFINE_MUTEX(register_count_mutex);
 static DEFINE_MUTEX(video_list_lock);
@@ -2225,6 +2228,22 @@ bool acpi_video_handles_brightness_key_p
 }
 EXPORT_SYMBOL(acpi_video_handles_brightness_key_presses);
 
+static ssize_t store_init_acpi_video(struct bus_type *bus,
+                                    const char *buf, size_t count)
+{
+	unsigned int val;
+	if (kstrtouint(buf, 10, &val) || val != 1)
+		return -EINVAL;
+	acpi_video_register();
+	return count;
+}
+
+static struct bus_attribute init_acpi_video_attr = {
+	.attr = {.name = "init_acpi_video", .mode = 0644},
+	.show = NULL,
+	.store = store_init_acpi_video,
+};
+
 /*
  * This is kind of nasty. Hardware using Intel chipsets may require
  * the video opregion code to be run first in order to initialise
@@ -2245,6 +2264,9 @@ static int __init acpi_video_init(void)
 	if (acpi_disabled)
 		return 0;
 
+	if (delay_init)
+		return bus_create_file(&acpi_bus_type, &init_acpi_video_attr);
+
 	if (intel_opregion_present())
 		return 0;
 
@@ -2253,6 +2275,9 @@ static int __init acpi_video_init(void)
 
 static void __exit acpi_video_exit(void)
 {
+	if (delay_init)
+		bus_remove_file(&acpi_bus_type, &init_acpi_video_attr);
+
 	acpi_video_detect_exit();
 	acpi_video_unregister();
 
